frontend/src/app/core/application.service.ts

```diff
@@ -0,0 +1,81 @@
+import { Injectable } from '@angular/core';
+import { Store } from '@ngrx/store';
+import type { AppState } from '../state/app.state';
+import {
+    selectExerciseId,
+    selectExerciseStateMode,
+    selectLastClientName,
+} from '../state/application/selectors/application.selectors';
+import { selectStateSnapshot } from '../state/get-state-snapshot';
+import { ExerciseService } from './exercise.service';
+import { TimeTravelService } from './time-travel.service';
+
+/**
+ * This service encapsulates the logic for switching between the live exercise, timeTravel and an empty state (e.g. on the landing page).
+ * Its API to switch between exercise state modes should be used instead of those in the {@link ExerciseService} and {@link TimeTravelService}.
+ */
+@Injectable({
+    providedIn: 'root',
+})
+export class ApplicationService {
+    constructor(
+        private readonly timeTravelService: TimeTravelService,
+        private readonly exerciseService: ExerciseService,
+        private readonly store: Store<AppState>
+    ) {}
+
+    /**
+     * A new mode must be set immediately after this function is called
+     */
+    private stopCurrentMode() {
+        const currentExerciseStateMode = selectStateSnapshot(
+            selectExerciseStateMode,
+            this.store
+        );
+        switch (currentExerciseStateMode) {
+            case 'exercise':
+                this.exerciseService.leaveExercise();
+                break;
+            case 'timeTravel':
+                this.timeTravelService.stopTimeTravel();
+                break;
+            case undefined:
+                break;
+            default:
+                assertExhaustiveness(currentExerciseStateMode);
+        }
+    }
+
+    public async joinExercise(
+        ...args: Parameters<typeof this.exerciseService.joinExercise>
+    ) {
+        this.stopCurrentMode();
+        return this.exerciseService.joinExercise(...args);
```

Comment:
During joining (async) and if `joinExercise` fails, the frontend is in an invalid state.
The same is true for `startTimeTravel`.